walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> spots = new ArrayList<TimeRange>();
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees());
+    ArrayList<Event> eventsList = new ArrayList<Event>(events);
+    long duration = request.getDuration();
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return spots;
+    }
+    
+    if (eventsList.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    /** Sorts the events list by start time. */
+    boolean sorted = true;
+    do {
+      for (int i = 0; i < eventsList.size() - 1; i++) {
+        if (TimeRange.ORDER_BY_START
+            .compare(eventsList.get(i).getWhen(), eventsList.get(i+1).getWhen()) > 0) {
+          Event tmp = eventsList.get(i);
+          eventsList.set(i, eventsList.get(i+1));
+          eventsList.set(i+1, tmp);
+          sorted = false;
+        }
+      }
+    } while (!sorted);
+
+    /** Deletes irrelevant events and events contained by other events. */
+    for (int i = 0; i < eventsList.size(); i++) {
+      boolean relevant = false;
+      int j = 0;
+      while (j < attendees.size() && (!relevant)) {
+        if (eventsList.get(i).getAttendees().contains(attendees.get(j))) {
+          relevant = true;
+        }
+        j++;
+      }
+      if (!relevant) {
+        eventsList.remove(i);
+        i--;
+      } else if ((i < eventsList.size() - 1) && 
+                 (eventsList.get(i).getWhen().contains(eventsList.get(i+1).getWhen()))) {
+        eventsList.remove(i+1);
+        i--;
+      }
+    }
```

Comment:
1. Since both sections require going through the list, by putting them together, the list is only visited once. Should I separate these two parts?

2. By sorting the eventsList, when looking for available time range, I only need to check whether the gap between two adjacent events is longer or equal to the duration of the request.

3. Although in the case above, the ith element will not be removed, it does not affect the final result as the gap between the ith and (i+1)th events is negative and thus shorter than the required duration. Maybe I should revise the comment.

Reply:
1. even you put them together, you visit the list twice. The time complexity doesn't change.

2. I didn't argue that you shouldn't use sort. I just think you need to both compare start and end.

3. If you argue that the result is correct in such a way, then what you want is just keep the time range sorted by: first key start, second key end. Then why do you want to perform removal rather than changing the sort comparator?

After thinking deeper, I wonder whether it's possible that one event contain the other? Since you only add a event when it doesn't intersect with others.

Reply:
I don't quite understand #3 here. Could you explain a little more?

It is possible that one event contains the other as they may belong to different people.

Reply:
> It is possible that one event contains the other as they may belong to different people.

Yes, but I mean those events that contains the attendees.

Reply:
for point 3, never mind. This is no longer a issue.
Since you have remove events that don't contain the attendees, the remained events won't intersect with each other. That means, you only need to compare the start.